algebird-core/src/main/scala/com/twitter/algebird/Scan.scala

```diff
@@ -0,0 +1,361 @@
+package com.twitter.algebird
+
+import scala.collection.compat._
+import scala.collection.generic.CanBuildFrom
+
+object Scan {
```

Comment:
How is this different than ` Aggregator[I,M[_]:SemigroupK, O]` . I think the motivation here is that you want `S` type to be a container so that `reduce` on `S` is collecting intermediate state. I believe that can be achieved by existing Aggregators by 1) making S a type constructor 2) making sure that S has an instance of [`SemigroupK`](https://typelevel.org/cats/typeclasses/semigroupk.html)

Reply:
@MansurAshraf First: great to be on a PR with you again! It's been a long time :-).

Second: the idea here is that this transforms streams to streams. whereas aggregators transform Sequences to single values. I would look at the `apply` method (starting on line 229 of `Stream.scala` as of this writing). The idea is that, for each input in the stream, you have one output in the stream. For example, using `directFreeScan` from `ScanTest`, something like `directFreeScan(List('a', 'b', 'c'))` would result in `List("a", "ab", "abc")`.

Does that make sense?

Reply:
hey @jeff-stripe , its been a while :-)

I will take a look at the Stream.scala but cursory look at `def fromAggregator[A, B, C](aggregator: Aggregator[A, B, C], initState: B): Aux[A, B, C]` gives me the impression that

```Scan[A,C] ~= Aggregator[A,S[_],(S[_],C)]```

Scan has a new function called `presentAndNextState` which return both S state type and final output type C. If you aggregator type was `Aggregator[A,S[_],(S[_],C)]`, the `present` function in existing aggregator will return the same value i think.

Reply:
@MansurAshraf In my example, the analogous `Aggregator` would just return `"abc"`, not `List("a", "ab", "abc")`. In other words, `Aggregator` is to the `reduce` method as `Scan` is to the `scanLeft` method.

Reply:
Mansur, to a first approximation, `Aggregator[A, _, B]` is `Iterator[A] => Option[B]`, `MonoidAggregator[A, _, B]` is `Iterator[A] => B` and `Scan[A, B]` is `Iterator[A] => Iterator[B]` (actually with an added law that the number of items in the result is exactly the same.

You could imagine another idea: `ExpandingScan[A, B]` that is `Iterator[A] => Iterator[B]` but without the law of the iterators being the same size (then we could do things like filter, or concatMap, etc...)

Does that help explain it at all?

Think of the `ExpandingScan` as the most general function you can do in a reducer in Hadoop.